### PR TITLE
WT-9120 Fix freeing the non obsolete prepared fast truncate pages

### PR DESCRIPTION
**Pull request by kommiharibabu**
Use the durable timestamp of the prepared truncate operation instead
of the commit timestamp to check the obsoleteness of the page.

### CODE REVIEW

File: test/suite/test_checkpoint06.py [hunk 1]
```diff
@@ -71,16 +80,37 @@ def test_rollback_truncation_in_checkpoint(self):
         start.set_key(5)
         end = self.session.open_cursor(self.uri)
         end.set_key(9995)
+        self.session.breakpoint()
```
Comment (quchenhao) [inline test/suite/test_checkpoint06.py:83]: This seems to be for debugging. Is it still needed?

Comment (kommiharibabu) [reply test/suite/test_checkpoint06.py]: No, removed. Thanks.

**Review by quchenhao — APPROVED** [review]
lgtm

**Review by sauclovian-wt — APPROVED** [review]
lgtm

**Review by keithbostic — APPROVED** [review]
lgtm

File: src/include/btree_inline.h [hunk 2]
```diff
@@ -1512,8 +1512,9 @@ __wt_page_del_active(WT_SESSION_IMPL *session, WT_REF *ref, bool visible_all)
     WT_ORDERED_READ(prepare_state, page_del->prepare_state);
     if (prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED)
         return (true);
-    return (visible_all ? !__wt_txn_visible_all(session, page_del->txnid, page_del->timestamp) :
-                          !__wt_txn_visible(session, page_del->txnid, page_del->timestamp));
+    return (visible_all ?
+        !__wt_txn_visible_all(session, page_del->txnid, page_del->durable_timestamp) :
+        !__wt_txn_visible(session, page_del->txnid, page_del->timestamp));
```
Comment (sauclovian-wt) [inline src/include/btree_inline.h:1517]: Should this be the durable timestamp in both cases or only for visible_all? If it was intentional I'm happy to go along, just want to make sure it's not an oversight.

**Comment by agorrod** [comment]
bq. Use the durable timestamp of the prepared truncate operation instead of the commit timestamp to check the obsoleteness of the page.

That doesn't sound right to me. The commit time needs to be less than or equal to the durable time, and we need to keep content available until the commit time has become obsolete, otherwise a reader won't be able to find something it should.

File: src/include/btree_inline.h (hunk 2 above)
Comment (kommiharibabu) [reply src/include/btree_inline.h:1517]: The visibility of a transaction is always compared against the `commit timestamp`, whereas the stability or obsoleteness of an update is decided against the `durable timestamp` of the operation. Due to this reason, I changed it to `durable timestamp` only for the `visible_all` case.

**Comment by kommiharibabu** [comment]
> Does this change mean content will remain "alive" for strictly longer than it otherwise could have, since the durable timestamp is newer than the commit timestamp and content is removed from oldest to newest?

Yes, it is correct that we retain the removed data that is no longer accessed by any transactions in the system for longer than is strictly required.

> It does feel like a place where we need to document the behavior very carefully. Since it's not intuitive, and I don't like places where the code needs to do something that isn't intuitive. To that end this change seems pretty opaque. Is there a way it could be wrapped in a function instead, that could have a BHC (big hunking comment) describing why the check is based on durable?

I am not sure that I can move the logic into a function as it is not only the place where we use the `durable timestamp`. All the callers of the `__wt_txn_visible_all` function already pass the `durable timestamp` except to check the readers of the metadata page. How about adding a comment on top of the `__wt_txn_visible_all` function describing how the `durable timestamp` is used to check for the global visibility and obsolete?

**Comment by sauclovian-wt** [comment]
In principle it's possible for something to want to know if something is _visible_ to everyone, as opposed to visible and also durable/stable. I don't think we have any such cases (I have visited most of the visible_all checks fairly recently due to WT-5927) and it wouldn't surprise me if none ever appear. (In general if we want to know if a value is visible, it's because we're deciding whether to observe it, and that's always in some specific context and not global; conversely visible_all checks are about whether any/all older values are obsolete, and that's about durability.) All the same, it might be a good idea in the long run to revise the terminology. Terminology reform is expensive; but it tends to pay for itself.

Not in this ticket though! :-)

**Comment by kommiharibabu** [comment]
@agorrod I added a comment to explain how the visible_all function is used. Please check and suggest if it can be enhanced.

File: src/include/txn_inline.h [hunk 3]
```diff
@@ -523,8 +523,10 @@ __txn_visible_all_id(WT_SESSION_IMPL *session, uint64_t id)
 
 /*
  * __wt_txn_visible_all --
- *     Check if a given transaction is "globally visible". This is, if all sessions in the system
- *     will see the transaction ID including the ID that belongs to a running checkpoint.
+ *     Check if a given time window is "globally visible" or "obsolete". This is, if all sessions in
+ *     the system will see the time window and it is durable. The timestamp passed to this function
+ *     is primarily the "durable timestamp" instead of a "commit timestamp" of an update/page to
+ *     check durability of an update/page whether it is "globally visible" or "obsolete".
```
Comment (agorrod) [inline src/include/txn_inline.h:529]: I'd suggest rewording to:

"Check whether a given time window is either globally visible or obsolete. For global visibility checks, the commit times are checked against the oldest possible readers in the system. If all possible readers could always see the time window - it is globally visible.
For obsolete checks callers should generally pass in the durable timestamp, since it is guaranteed to be newer than or equal to the commit time, and content needs to be retained (not become obsolete) until both the commit and durable times are obsolete. If the commit time is used for this check, it's possible that a transaction is committed with a durable time and made obsolete before it can be included in a checkpoint - which leads to bugs in checkpoint correctness."

Comment (kommiharibabu) [reply src/include/txn_inline.h]: Thanks. Fixed as you suggested.

**Review by agorrod — APPROVED** [review]
LGTM - sorry for the delay.
